### PR TITLE
[SPARK-47272][SS] Add MapState implementation for State API v2. 

### PR DESCRIPTION
**Pull request by jingz-db**
### What changes were proposed in this pull request?

This PR adds changes for MapState implementation in State Api v2. This implementation adds a new encoder/decoder to encode grouping key and user key into a composite key to be put into RocksDB so that we could retrieve key-value pair by user specified user key by one rocksdb get.

### Why are the changes needed?

These changes are needed to support map values in the State Store. The changes are part of the work around adding new stateful streaming operator for arbitrary state mgmt that provides a bunch of new features listed in the SPIP JIRA here - https://issues.apache.org/jira/browse/SPARK-45939

### Does this PR introduce _any_ user-facing change?

Yes
This PR introduces a new state type (MapState) that users can use in their Spark streaming queries.

### How was this patch tested?

Unit tests in `TransforWithMapStateSuite`.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:64]: This is an ugly implementation. We may consider use similar ways as `useMultipleValuesPerKey`. But we may get lost in `TransformWithStateExec` in calling `createAndInit`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:33]: We can also pass in a two-column key schema to `PrefixKeyScanStateEncoder`.

**Comment by jingz-db** [comment]
Thanks Eric for reviews on my old PR. I've resolved them and incorporated in this one already.

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/MapState.scala
Comment (anishshri-db) [inline sql/api/src/main/scala/org/apache/spark/sql/streaming/MapState.scala:53]: Can we rename this to `clear` ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:71]: Lets serialize using Spark SQL here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:32]: Can we not encode the grouping key and user key together here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:87]: Not sure I understand why we need to update the store providers themselves ? Can't we just use an updated schema that encodes the required information within the key

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:33]: Yes - could we please do this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:80]: Could we reuse this instead of creating new one each time ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:83]: Same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:88]: Same here as well

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:98]: Same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:54]: To be consistent, I think we can return `null` here similar to other state types

```
null.asInstanceOf[V]
```

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:196]: Can we test for `clear` too ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:104]: Nit: `non-existing` or `absent`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:34]: You probably need to merge back from master to update this ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:28]: Could we also add pure unit tests similar to ValueStateSuite - https://src.dev.databricks.com/databricks/runtime@master/-/blob/sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala?subtree=true ? Thx

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:52]: Its ok to reuse I think. As far as the user is concerned, they are writing a value out. So should be ok to reuse

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:65]: I guess we can remove the `try catch` block now ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:117]: I guess you can just directly call `keyOption.get` here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:133]: Hmm - do we really need these singleton objects ? Could we just call `new CompositeKeyStateEncoder` instead ?
cc - @HeartSaVioR - is this a preferred pattern for Spark ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:98]: Lets pass the schema as an argument to make this more generic ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:90]: Could we move this to a new file please ? Thx

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:124]: Compiler will complain in the next line:
`keySerializer.apply(groupingKey)` where groupingKey will be of `Any` type if we directly call `keyOption.get`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:198]: Same here

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:126]: Can we add a test for batch query using `mapState` ?

**Comment by anishshri-db** [comment]
@jingz-db - test failure seems related ?

**Comment by jingz-db** [comment]
> @jingz-db - test failure seems related ?

Weirdly is passing locally. Let me resolve your comments and retrigger the CI and see if it still fails. Thanks for the review!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:137]: I was following Bhuwan's style in the base class. Maybe I am missing something but did not find anything useful in the [style guide](https://github.com/databricks/scala-style-guide).

**Comment by anishshri-db** [comment]
@jingz-db - could you please fix this style error ?

```
[info] compiling 1 Scala source to /home/runner/work/spark/spark/tools/target/scala-2.13/classes ...
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:195:0: Whitespace at end of line
[info]   Compilation completed in 14.316s.
```

**Review by anishshri-db — APPROVED** [review]
LGTM ! pending CI completion with success

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:38]: nit: indentation

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:99]: nit: `{ e =>` rather than `(e => {`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:100]: Could we verify the exception with checkError?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:144]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateSuite.scala:143]: ditto

**Review by HeartSaVioR — COMMENTED** [review]
Mostly looks good. Left some comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala]: Hi @anishshri-db, need your input on this: Do we want to return `Map` type or `Iterator` type for `getMap` function?
Talked with Jungtaek on Slack, if we decide to return `Map` type, we'll probably need to materialize the map and copy everything in map into memory (because we reuse UnsafeRow in StateTypeEncoder). So Jungtaek is concerning about the case where we have a large map. I also feel like returning Iterator type makes more sense, because for ListState we also return Iterator for get list function.

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala]: Yea sure - lets use the iterator approach. Thx

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MapStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MapStateSuite.scala:70]: Maybe just do this in setup (before) or teardown (after) in StateVariableSuiteBase, to ensure the cleanup is guaranteed to be done - as failure in test A won't clean up the thread local and introduce another failure. We do like to avoid cascading failures.

Comment (jingz-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MapStateSuite.scala]: Done! Thanks Jungtaek!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:81]: maybe final nit: your previous code just works and it seems much simpler. We just don't call `.toMap` and done. 

Also maybe the method name to remove the map from the name? 

* `getIterator` - consistent with getKeys/getValues
* `iterator` - consistent with Map collection. need to change other methods as well, e.g. getKeys to keys, getValues to values
* `get()` - consistent with other type of state. The value of origin type is retrieved with get() consistently.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala:81]: It's OK to defer the change of method name in follow up PR if we want to have a time to figure out the best name. The first one is something good to do this before merging the code.

Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImpl.scala]: Thanks Jungtaek! As discussed, I changed `getMap()` name to `iterator` and pair it with `getKeys()`, `getValues()` to `keys()` and `values()` accordingly.

**Comment by jingz-db** [comment]
Test suite failure seems unrelated in pyspark.

**Comment by HeartSaVioR** [comment]
Yeah, doesn't look to be related. I'm ignoring the CI error.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
